ide/libs.git/src/org/netbeans/libs/git/jgit/commands/ListBranchCommand.java

```diff
@@ -41,61 +44,58 @@ public class ListBranchCommand extends GitCommand {
     private final boolean all;
     private Map<String, GitBranch> branches;
 
-    public ListBranchCommand (Repository repository, GitClassFactory gitFactory, boolean all, ProgressMonitor monitor) {
+    public ListBranchCommand(Repository repository, GitClassFactory gitFactory, boolean all, ProgressMonitor monitor) {
         super(repository, gitFactory, monitor);
         this.all = all;
     }
 
     @Override
-    protected void run () throws GitException {
+    protected void run() throws GitException {
         Repository repository = getRepository();
-        Map<String, Ref> refs;
+        Map<String, GitBranch> map = new LinkedHashMap<>();
         try {
-            refs = repository.getAllRefs();
-        } catch (IllegalArgumentException ex) {
-            throw new GitException("Corrupted repository metadata at " + repository.getWorkTree().getAbsolutePath(), ex); //NOI18N
-        }
-        Ref head = refs.get(Constants.HEAD);
-        branches = new LinkedHashMap<String, GitBranch>();
-        Config cfg = repository.getConfig();
-        if (head != null) {
-            String current = head.getLeaf().getName();
-            if (current.equals(Constants.HEAD)) {
-                String name = GitBranch.NO_BRANCH;
-                branches.put(name, getClassFactory().createBranch(name, false, true, head.getLeaf().getObjectId()));
+            RefDatabase db = repository.getRefDatabase();
+            Ref head = db.exactRef(Constants.HEAD);
+            if (head != null) {
+                String current = head.getLeaf().getName();
+                if (current.equals(Constants.HEAD)) {
+                    String name = GitBranch.NO_BRANCH;
+                    map.put(name, getClassFactory().createBranch(name, false, true, head.getLeaf().getObjectId()));
+                }
+                map.putAll(getRefs(db.getRefsByPrefix(Constants.R_HEADS), false, current));
             }
-            branches.putAll(getRefs(refs.values(), Constants.R_HEADS, false, current, cfg));
-        }
-        Map<String, GitBranch> allBranches = getRefs(refs.values(), Constants.R_REMOTES, true, null, cfg);
-        allBranches.putAll(branches);
-        setupTracking(branches, allBranches, repository.getConfig());
-        if (all) {
-            branches.putAll(allBranches);
+            Map<String, GitBranch> allBranches = getRefs(db.getRefsByPrefix(Constants.R_REMOTES), true, null);
+            allBranches.putAll(map);
+            setupTracking(map, allBranches, repository.getConfig());
+            if (all) {
+                map.putAll(allBranches);
+            }
+            branches = Collections.unmodifiableMap(map);
```

Comment:
yeah, but if someone modifies a returned collection it is usually not intended and/or a bug.

Imagine someone passing the `ListBranchCommand` instance around, to UI or utilities, and some place is mutating the object by accident after calling a getter. This could not only break the client code, but also NB due to shared state.

Reply:
What you say is, that I have to copy defensively _every_ List I get from an API because it might suddenly turn read-only. I would expect a command to get me "my" list and if that is supported by the implementation, then this was made a promise.

Reply:
Forget it, the Commands are not exported

Reply:
if you want to mutate it yes. You shouldn't call a _getter_ and then mutate the collection. NB has also a hint for this (which is a bit buggy atm) which will warn if an object leaks mutable collections. This would ask for concurrent modification exceptions otherwise (best case scenario, worse is if no exceptions occur).

The jgit API used in this class for example does the same: `db.getRefsByPrefix(..)` returns an unmodifiable collection which is then added to a mutable Collection, to be modified and then frozen again before its returned via a getter.

Reply:
i mean, there is always the option to do this "old school":
getter...
```java
return new HashMap(internalMap);
```

Reply:
But this PR was also not about collections, so if you still prefer to keep the old behavior I can revert the change.

Reply:
A cleanup should IMHO not change behavior and that is the case now. I would not veto this, but I'd like to see the original, outside visible, behavior restored.

Reply:
reverted it, diff: https://github.com/apache/netbeans/compare/124c4dba530535949762d72c3faac97c5b8f2a98..e02c519361f9412200899727aea9e7283d3e6702